File: webapp/backend/routers/predicates.py
from fastapi import APIRouter, Query

router = APIRouter(prefix="/predicates", tags=["predicates"])


@router.get("/summary")
def summary():
    from kg.queries.predicates import predicate_summary
    df = predicate_summary()
    return {"data": df.to_dict(orient="records")}


@router.get("/pairs")
def pairs(
    concept_type: str = Query("all", regex="^(all|species|genus|feature)$"),
    limit: int = Query(50, ge=1, le=500),
):
    from kg.queries.predicates import part_of_pairs
    df = part_of_pairs(concept_type, limit)
    if df.empty:
        return {"data": [], "total": 0}
    return {"data": df.to_dict(orient="records"), "total": len(df)}


@router.get("/graph")
def graph():
    from kg.queries.predicates import predicate_graph
    return predicate_graph()

File: webapp/frontend/src/api/predicates.ts
import { fetchJson } from "./client";
import type {
  PredicateSummary,
  PredicatePair,
  PredicateGraphNode,
  PredicateGraphEdge,
} from "../types";

export async function getPredicateSummary(): Promise<{ data: PredicateSummary[] }> {
  return fetchJson("/predicates/summary");
}

export async function getPredicatePairs(
  conceptType: string = "all",
  limit: number = 50,
): Promise<{ data: PredicatePair[]; total: number }> {
  return fetchJson(`/predicates/pairs?concept_type=${conceptType}&limit=${limit}`);
}

export async function getPredicateGraph(): Promise<{
  nodes: PredicateGraphNode[];
  edges: PredicateGraphEdge[];
}> {
  return fetchJson("/predicates/graph");
}
